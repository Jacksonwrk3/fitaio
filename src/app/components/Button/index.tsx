"use client";

/**
 * @description Type for the Button component's props
 * @property {React.ReactNode} children - The content inside the button
 * @property {(e?: React.MouseEvent) => void} onClick - The function to be called when the button is clicked
 * @property {string} variant - The variant of the button
 * @property {string} width - The width of the button
 * @property {string} icon - The icon to be displayed
 */
type ButtonProps = {
  children: React.ReactNode;
  onClick?: (e?: React.MouseEvent) => void;
  variant?: "primary" | "outlined";
  width?: "full" | "auto";
  icon?: "string";
  disabled: boolean;
};

/**
 * @component Button
 * @type {React.FC<ButtonProps>}
 * @dev - Width defaults to auto
 * - Width[full] - Sets the width to 100%
 * @dev - Variant defaults to primary
 * @returns {JSX.Element} The rendered button element
 */
const Button: React.FC<ButtonProps> = ({
  children,
  onClick,
  variant = "primary",
  width = "auto",
  icon,
  disabled = false,
}) => {
  const primaryClasses = "bg-blue-600 text-white hover:bg-blue-700";
  const secondaryClasses =
    "bg-white hover:bg-gray-100 border-grayPrimary border text-black";
  const widthClasses = width === "full" ? "w-full" : "w-auto";
  return (
    <button
      onClick={onClick}
      disabled={disabled}
      className={`${
        variant === "primary" ? primaryClasses : secondaryClasses
      }  ${widthClasses} px-4 py-3 rounded`}
    >
      {children}
    </button>
  );
};

export default Button;
