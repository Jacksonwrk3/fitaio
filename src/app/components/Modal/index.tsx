import { useEffect } from "react";
import { createPortal } from "react-dom";

/**
 * @interface ModalProps
 * @property {React.ReactNode} children - Modal content
 * @property {boolean} isOpen - Determines if the modal is displayed or not
 * @property {() => void} onClose - Callback function that closes the modal
 * @property {string} target - The element id we are attaching our modal to
 */
interface ModalProps {
  children: React.ReactNode;
  isOpen: boolean;
  onClose: () => void;
  target: string;
}

/**
 * @component Modal
 * @description A reusuable modal component created using React.createPortal
 * @interface {ModalProps}
 * @returns {JSX.Element} The rendered Modal element
 */
const Modal: React.FC<ModalProps> = ({ isOpen, onClose, children, target }) => {
  useEffect(() => {
    //Disable scrolling when modal is open
    if (isOpen) {
      document.body.style.overflowY = "hidden";
    }

    // Cleanup function to enable scrolling when the modal is closed
    return () => {
      document.body.style.overflowY = "auto";
    };
  }, [isOpen]);
  //If is Open is false return null
  if (!isOpen) {
    return null;
  }
  //Retrieves the HTML element with the specified id (target) ans cast its type as HTMLElement.
  //If it doesn't exist, modal root will be null
  const modalRoot = document.getElementById(target) as HTMLElement;

  //If modal root is null then we return null and console.error
  if (!modalRoot) {
    console.error(`Modal root not found: ${target}`);
    return null;
  }
  return createPortal(
    <div
      className="w-screen h-screen absolute top-0 right-0 flex items-center justify-center bg-black bg-opacity-80"
      onClick={onClose}
    >
      <div className="max-w-3xl  w-[90%] border-1 bg-white border-grayPrimary rounded relative ">
        <button className="absolute right-8 top-4" onClick={onClose}>
          x
        </button>
        <div className="overflow-y-auto h-[90vh] max-h-[567px]">{children}</div>
      </div>
    </div>,
    modalRoot
  );
};

export default Modal;
