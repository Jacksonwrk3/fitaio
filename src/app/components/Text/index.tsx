import { create } from "domain";
import React, { createElement } from "react";
import { text } from "stream/consumers";
/**
 * @interface TextProps
 * @property {string} children - Text to be displayed
 * @property {string} as - Which html element the component will represent
 * @property {string} fontSize = One of the tailwind css font size classes
 * @property {string} casing - Text-transform property
 */

interface TextProps {
  fontSize?:
    | "text-xs"
    | "text-sm"
    | "text-base"
    | "text-lg"
    | "text-xl"
    | "text-2xl"
    | "text-3xl"
    | "text-4xl"
    | "text-5xl"
    | "text-6xl"
    | "text-7xl"
    | "text-8xl"
    | "text-9xl";
  children?: string;
  as?:
    | "h1"
    | "h2"
    | "h3"
    | "h4"
    | "h5"
    | "h6"
    | "p"
    | "span"
    | "strong"
    | "em"
    | "small"
    | "mark"
    | "del"
    | "ins";
  casing?: "uppercase" | "lowercase" | "capitalize" | "normal-case";
  align?:
    | "text-left"
    | "text-center"
    | "text-right"
    | "text-justify"
    | "text-start"
    | "text-end";
  color?: "white" | "red";
}

/**
 * @description Text Component
 * @param {string} children - Text to be displayed
 * @param {string} [as="p"] - Which HTML element the component will represent
 *    - Defaults to "p".
 * @param {string} [fontSize="text-base"] - Tailwind CSS font size class
 *    - Options: "text-xs", "text-sm", "text-base", "text-lg", "text-xl", "text-2xl", "text-3xl", "text-4xl", "text-5xl", "text-6xl", "text-7xl", "text-8xl", "text-9xl"
 *    - Defaults to "text-base".
 *    - See the [Font Size Documentation](https://tailwindcss.com/docs/font-size).
 * @param {string} [casing="normal-case"] - Text-transform property
 *    - Options: "uppercase", "lowercase", "capitalize", "normal-case".
 *    - Defaults to "normal-case".
 *    - See the [Text Transform Documentation](https://tailwindcss.com/docs/text-transform).
 * @param {string} [align="text-left"] - Text alignment class
 *    - Options: "text-left", "text-center", "text-right", "text-justify", "text-start", "text-end".
 *    - Defaults to "text-left".
 *    - See the [Text Alignment Documentation](https://tailwindcss.com/docs/text-align).
 * @param {string}
 */

const Text: React.FC<TextProps> = ({
  children,
  as = "p",
  fontSize = "text-base",
  casing = "normal-case",
  align = "text-left",
  color,
}) => {
  let textColor;
  switch (as) {
    case "h1":
    case "h2":
    case "h3":
    case "h4":
    case "h5":
    case "h6":
      textColor = "text-dark";
      break;
    case "p":
    case "span":
    case "strong":
    case "em":
    case "small":
    case "mark":
    case "del":
    case "ins":
      textColor = "text-primary";
      break;
    default:
      textColor = "text-primary";
      break;
  }
  switch (color) {
    case "white":
      textColor = "text-white";
      break;
    case "red":
      textColor = "text-red-500";
  }
  const className = `${textColor} ${fontSize} ${casing} ${align}`;
  return createElement(as, { className: className }, children);
};

export default Text;
