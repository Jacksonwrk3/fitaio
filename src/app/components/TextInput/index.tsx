import { useState, useEffect } from "react";
import TextInputProps from "./TextInput.types";

/**
 * @description A TextInput component that renders a text input field, either of type "text" or "password".
 * It includes the ability to toggle the visibility of the password when the type is set to "password".
 * The component also requires either an `onChange` or `onBlur` handler (or both).
 * @param {TextInputProps} props - The props for the TextInput component
 * @returns {JSX.Element} The rendered TextInput component
 */
const TextInput: React.FC<TextInputProps> = ({
  id,
  placeholder,
  onChange,
  onBlur,
  value,
  type = "text",
}) => {
  const [isVisible, setIsVisible] = useState(false);

  /**
   * Toggles visibility of the password input (from password to text and vice versa).
   */
  const toggleVisibility = () => {
    setIsVisible((prev) => !prev);
  };

  // Effect hook to set the visibility based on the input type
  useEffect(() => {
    if (type === "text") {
      setIsVisible(true);
    }
  }, [type]); // Only re-run when type changes

  return (
    <div className="relative flex items-center border border-grayPrimary rounded">
      <input
        type={isVisible ? "text" : "password"} // Toggle between text and password visibility
        id={id}
        name={id}
        className="py-2 pl-4 w-full focus:outline-none focus:ring-0 box-border focus:border focus:border-primary-500 focus:bg-primary-100"
        placeholder={placeholder}
        onChange={(e) => onChange && onChange(e)} // Only call onChange if it's provided
        onBlur={(e) => onBlur && onBlur(e)} // Only call onBlur if it's provided
        value={value} // Controlled input value
      />
      {type === "password" && (
        <div
          className="text-primary-500 absolute right-0 mr-4 cursor-pointer"
          onClick={toggleVisibility}
        >
          {isVisible ? "Hide" : "Show"} {/* Toggle visibility text */}
        </div>
      )}
    </div>
  );
};

export default TextInput;
