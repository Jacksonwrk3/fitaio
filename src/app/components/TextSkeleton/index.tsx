/**
 * @description Fallback component for when aysnchronous text is loading
 */
const TextSkeleton = () => {
  return <div className="w-full animate-pulse p-2 bg-loadinGray "></div>;
};

export default TextSkeleton;
