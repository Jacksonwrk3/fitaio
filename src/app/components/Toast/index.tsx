import React from "react";

/**
 * @description ToastProps
 * @property {"success" | "warning" | "error" | "info"} status
 * @property {string} title
 * @property {string} description
 */
interface ToastProps {
  status: "success" | "warning" | "error" | "info";
  title: string;
  description?: string;
}

const Toast: React.FC<ToastProps> = () => {};

export default Toast;
