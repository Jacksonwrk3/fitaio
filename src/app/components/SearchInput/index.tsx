import React, { useState, Suspense } from "react";
import TextInput from "../TextInput";
import TextSkeleton from "../TextSkeleton";
import Loading from "./loading";
/**
 * @interface SearchInputProps
 * @property {string} id - ID for TextInput
 * @property {(e: React.ChangeEvent<HTMLInputElement>) => void | Promise<void>} - Function that runs when TextInput changes
 * @property {string[]} items - Auto-complete items
 * @property {value} value - Value of the inputs entered into the text box
 */
interface SearchInputProps {
  id: string;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void | Promise<void>;
  items: string[];
  value: string;
}

/**
 * @description Search Bar that gives suggestions as you type
 * @param {string} id - ID for TextInput
 * @param {(e: React.ChangeEvent<HTMLInputElement>) => void | Promise<void>} onChange - Function that runs when TextInput changes
 * @param {string[]} items - Auto-complete items
 * @param {value} value - Value of the inputs entered into the text box
 */
const SearchInput: React.FC<SearchInputProps> = ({
  id,
  onChange,
  items,
  value,
}) => {
  const [error, setError] = useState(null);
  return (
    <div className="flex flex-col space-y-1">
      <TextInput id={id} onChange={onChange} value={value} />
      {items ? (
        <div className="border flex flex-col space-y-1 p-3 border-loadingGray w-full  rounded">
          {/* List items. If items aren't done fetching load fallback UI */}
          <Suspense fallback={<Loading />}>
            <div>Push up</div>
            <div>Sit up</div>
            <div>Back row</div>
            <div>Tricip pull down</div>
            <div>Curls</div>
          </Suspense>
        </div>
      ) : null}
    </div>
  );
};

export default SearchInput;
