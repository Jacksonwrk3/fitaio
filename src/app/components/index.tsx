import Button from "./Button";
import TextInput from "./TextInput";
import Modal from "./Modal";
import Text from "./Text";
import Toast from "./Toast";
import SearchInput from "./SearchInput";
import TextSkeleton from "./TextSkeleton";
export { Button, TextInput, Modal, Toast, Text, SearchInput, TextSkeleton };
