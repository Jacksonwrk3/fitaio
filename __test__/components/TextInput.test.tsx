import { render, screen, fireEvent } from "@testing-library/react";
import { TextInput } from "@/app/components";
import "@testing-library/jest-dom"; // for the extended matchers

describe("TextInput Component", () => {
  const mockOnChange = jest.fn();
  const mockOnBlur = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render with the provided value and placeholder", () => {
    render(
      <TextInput
        id="test-input"
        value="Test Value"
        onChange={mockOnChange}
        onBlur={mockOnBlur}
        placeholder="Enter text"
      />
    );

    const inputElement = screen.getByPlaceholderText("Enter text");
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveValue("Test Value");
  });

  it("Should call onChange when input value changes", () => {
    render(
      <TextInput
        id="test-input"
        placeholder="Enter text"
        onChange={mockOnChange}
        value=""
      />
    );

    const input = screen.getByPlaceholderText("Enter text");

    // Simulate typing into the input field
    fireEvent.change(input, { target: { value: "New Value" } });

    // Assert that onChange was called
    expect(mockOnChange).toHaveBeenCalledTimes(1);
  });

  it("should call onBlur when input loses focus", () => {
    render(
      <TextInput
        id="test-input"
        value="Test Value"
        onChange={mockOnChange}
        onBlur={mockOnBlur}
        placeholder="Enter text"
      />
    );

    const inputElement = screen.getByPlaceholderText("Enter text");
    fireEvent.blur(inputElement);

    expect(mockOnBlur).toHaveBeenCalledTimes(1);
  });

  it('should toggle visibility when the "Show/Hide" button is clicked for password input', () => {
    render(
      <TextInput
        id="test-input"
        value="Password"
        onChange={mockOnChange}
        onBlur={mockOnBlur}
        placeholder="Enter password"
        type="password"
      />
    );

    const showHideButton = screen.getByText("Show");
    expect(showHideButton).toBeInTheDocument();

    // Initially it should be password type
    const inputElement = screen.getByPlaceholderText("Enter password");
    expect(inputElement).toHaveAttribute("type", "password");

    // Click to toggle visibility
    fireEvent.click(showHideButton);

    // After click, the input type should be text
    expect(inputElement).toHaveAttribute("type", "text");
    expect(screen.getByText("Hide")).toBeInTheDocument();
  });

  it("should match snapshot for text input", () => {
    const { asFragment } = render(
      <TextInput
        id="test-input"
        value="Text"
        onChange={mockOnChange}
        onBlur={mockOnBlur}
        placeholder="Enter text"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render a password field and show/hide the password visibility toggle", () => {
    render(
      <TextInput
        id="password-input"
        value="Password123"
        onChange={mockOnChange}
        onBlur={mockOnBlur}
        placeholder="Enter password"
        type="password"
      />
    );

    // The input should initially be of type password
    const inputElement = screen.getByPlaceholderText("Enter password");
    expect(inputElement).toHaveAttribute("type", "password");

    // "Show" text should be visible initially
    expect(screen.getByText("Show")).toBeInTheDocument();
  });
});
