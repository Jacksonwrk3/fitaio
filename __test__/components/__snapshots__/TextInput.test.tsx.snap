// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Component should match snapshot for text input 1`] = `
<DocumentFragment>
  <div
    class="relative flex items-center border border-grayPrimary rounded"
  >
    <input
      class="py-2 pl-4 w-full focus:outline-none focus:ring-0 box-border focus:border focus:border-primary-500 focus:bg-primary-100"
      id="test-input"
      name="test-input"
      placeholder="Enter text"
      type="text"
      value="Text"
    />
  </div>
</DocumentFragment>
`;
